components/locid/src/locale.rs

```diff
@@ -149,6 +149,55 @@ impl Locale {
         self.extensions.unicode.keywords.get(key)
     }
 
+    /// Mutating setter method for Unicode extensions.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu::locid::{extensions::unicode::Key, locale, unicode_ext_key};
+    ///
+    /// const KEY: Key = unicode_ext_key!("hc");
+    ///
+    /// let mut loc = locale!("en-US");
+    /// loc.set_unicode_extension(KEY, "h12".parse().expect("valid subtag"));
+    /// assert_eq!(
+    ///     loc,
+    ///     "en-US-u-hc-h12"
+    /// );
+    /// ```
+    pub fn set_unicode_extension(
+        &mut self,
+        key: extensions::unicode::Key,
+        value: extensions::unicode::Value,
+    ) {
+        self.extensions.unicode.keywords.set(key, value);
+    }
+
+    /// Chainable setter method for Unicode extensions.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu::locid::{extensions::unicode::Key, locale, unicode_ext_key};
+    ///
+    /// const KEY: Key = unicode_ext_key!("hc");
+    ///
+    /// let loc = locale!("en-US")
+    ///     .with_unicode_extension(KEY, "h12".parse().expect("valid subtag"));
+    /// assert_eq!(
+    ///     loc,
+    ///     "en-US-u-hc-h12"
+    /// );
+    /// ```
+    pub fn with_unicode_extension(
+        mut self,
+        key: extensions::unicode::Key,
+        value: extensions::unicode::Value,
+    ) -> Self {
+        self.extensions.unicode.keywords.set(key, value);
+        self
+    }
```

Comment:
This seems like it introduces a builder model. I'm slightly opposed to it here, but less so than in other places since it doesn't lead to inconsistent mid-point state. But I don't think it's that much easier to read than:

```rust
let mut loc = locale!("en-US");
loc.extensions.unicode.set(KEY, "h12".parse().expect(...));
```